cel/optimizer.go

```diff
@@ -388,3 +343,66 @@ func (opt *optimizerExprFactory) NewStruct(typeName string, fields []ast.EntryEx
 func (opt *optimizerExprFactory) NewStructField(field string, value ast.Expr, isOptional bool) ast.EntryExpr {
 	return opt.fac.NewStructField(opt.nextID(), field, value, isOptional)
 }
+
+// UpdateExpr updates the previous expression with the updated content while preserving macro metadata.
+func (opt *optimizerExprFactory) UpdateExpr(prev, updated ast.Expr) {
+	// Update the expression
+	prev.SetKindCase(updated)
+	if len(opt.sourceInfo.MacroCalls()) == 0 {
+		return
+	}
+
+	// prev is not macro, updated is not macro
+	// prev is macro, updated is not macro
+	// prev is macro, updated is macro
+	// prev is not macro, updated is macro
+
+	// Determine whether the previous expression was a macro.
+	_, prevIsMacro := opt.sourceInfo.GetMacroCall(prev.ID())
+
+	// Determine whether the updated expression was a macro.
+	updatedMacro, updatedIsMacro := opt.sourceInfo.GetMacroCall(updated.ID())
+
+	if updatedIsMacro {
+		// If the updated call was a macro, then updated id maps to prev id,
+		// and the updated macro moves into the prev id slot.
+		opt.sourceInfo.ClearMacroCall(updated.ID())
+		opt.sourceInfo.SetMacroCall(prev.ID(), updatedMacro)
+	} else if prevIsMacro {
+		// Otherwise if the previous expr was a macro, but is no longer, clear
+		// the macro reference.
+		opt.sourceInfo.ClearMacroCall(prev.ID())
+	}
+
+	// Punch holes in the updated value where macros references exist.
+	macroExpr := opt.fac.CopyExpr(prev)
+	macroRefVisitor := ast.NewExprVisitor(func(e ast.Expr) {
+		if _, exists := opt.sourceInfo.GetMacroCall(e.ID()); exists {
+			e.SetKindCase(nil)
+		}
+	})
+	ast.PostOrderVisit(macroExpr, macroRefVisitor)
+
+	// Update any references to the expression within a macro
+	macroVisitor := ast.NewExprVisitor(func(call ast.Expr) {
+		if call.ID() == prev.ID() {
+			call.SetKindCase(opt.fac.CopyExpr(macroExpr))
+		}
+		if call.ID() == updated.ID() {
+			if updatedIsMacro {
+				call.SetKindCase(nil)
+			} else {
+				call.SetKindCase(opt.fac.CopyExpr(macroExpr))
+			}
+			call.RenumberIDs(func(id int64) int64 {
+				if id == updated.ID() {
+					return prev.ID()
+				}
+				return id
+			})
```

Comment:
I actually don't fully understand why this step is needed. My take on this is that we're accounting for orphaned macro expr nodes after update, then normalizing the IDs afterwards. Some comment may help.

Reply:
I've added more comments